### PR TITLE
Better accessibility support#82

### LINKED ISSUE #82: better accessibility support
The following features should be implemented for better accessibility and 508 compliance:

- [x] https://github.com/d3plus/d3plus-shape/issues/78
- [x] https://github.com/d3plus/d3plus-shape/issues/79
- [x] https://github.com/d3plus/d3plus-text/issues/89
- [x] add a `<title id="${uuid}-title">` tag containing `this._title || "D3plus Visualization"` ([docs](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/title) on SVG titles, code will probably need to be inserted [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L418))
- [x] set `ariaHidden: true` on the visible `TextBox` chart title (should be able to add to titleConfig [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L187-L192))
- [x] add a `<desc id="${uuid}-desc">` tag mapped to a new `desc` method and `this._desc` (https://developer.mozilla.org/en-US/docs/Web/SVG/Element/desc) on SVG titles, code will probably need to be inserted [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L418))
- [x] add `aria-labelledby="${this._uuid}-title ${this._uuid}-desc"` to the `<svg>` container (SVG is modified [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L426-L428))
- [x] add `role="img"` to the `<svg>` container (SVG is modified [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L426-L428))
- [x] add `role="presentation"` to all shapes (can probably be added to the shapeConfig [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L150-L180))
- [x] set `ariaLabel` for all shapes to the `drawLabel` (to be overridden by charts as necessary) (can probably be added to the shapeConfig [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L150-L180))
- [x] set `ariaLabel` for all legend shapes to the `legendLabel` (can probably be added to the legendConfig [here](https://github.com/d3plus/d3plus-viz/blob/master/src/Viz.js#L100-L109))

### PR DESCRIPTION
**Pull request by rbaheti**
### Description

- Added support for SVG title and desc tags, TextBox,  SVG container, all shapes and legends.
 closes #82

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have documented all new methods.
- [ ] I have written tests for all new methods/functionality.
- [ ] I have written examples for all new methods/functionality.

### CODE REVIEW

File: src/Viz.js [hunk 1]
```diff
@@ -660,6 +672,16 @@ If *data* is not specified, this method returns the current primary data array,
...
   }
 
+  /**
+      @memberof Viz
+      @desc If *value* is specified, sets the description accessor to the specified function or string and returns the current class instance.
+      @param {Function|String} [*value*]
+      @chainable
+  */
+  desc(_) {
+    return arguments.length ? (this._desc = typeof _ === "function" ? _ : constant(_), this) : this._desc;
```
Comment (davelandry) [inline src/Viz.js]: `this._desc` should only ever be a *String*, so remove the function code from here

File: src/Viz.js [hunk 2]
```diff
@@ -148,6 +149,7 @@ export default class Viz extends BaseClass {
     this._shape = constant("Rect");
     this._shapes = [];
     this._shapeConfig = {
+      ariaLabel: this._drawLabel,
```
Comment (davelandry) [inline src/Viz.js]: looking at the chrome element inspector, it doesn't appear this is getting applied to the shapes...

File: src/Viz.js [hunk 3]
```diff
@@ -98,6 +98,7 @@ export default class Viz extends BaseClass {
     this._groupBy = [accessor("id")];
     this._legend = true;
     this._legendConfig = {
+      ariaLabel: legendLabel.bind(this),
```
Comment (davelandry) [inline src/Viz.js]: looking at the chrome element inspector, it doesn't appear this is getting applied to the legend shapes. try moving this into the nested `shapeConfig` a few lines below, as `ariaLabel` is a property of the legend shapes, not the legend itself.

**Comment by rbaheti** [comment]
Thanks Dave for reviewing my PR.

1. Oops, misunderstood your previous comment on .desc()! Will make the
changes as requested above.
2. Cool, will make changes accordingly.

Thanks,
Rashmi

File: src/Viz.js [hunk 4]
```diff
@@ -660,6 +674,16 @@ If *data* is not specified, this method returns the current primary data array,
...
   }
 
+  /**
+      @memberof Viz
+      @desc If *value* is specified, sets the description accessor to the specified string and returns the current class instance.
+      @param {String} [*value*]
+      @chainable
+  */
+  desc(_) {
+    return arguments.length ? (this._svgDesc =  _ ) : this._svgDesc;
```
Comment (davelandry) [inline src/Viz.js]: This line is not correct, and causes an error when trying to use the method. You are assigning the `_svgDesc`, but not returning anything (so it's not chainable). Referencing the method above, it should be this:
```js
return arguments.length ? (this._svgDesc =  _, this) : this._svgDesc;
```

Please make sure that you have a javascript linter installed in your editor. That was the first thing to tip me off that this is where there are errors (`npm test` also fails).

What my Atom.io linting looks like:

screen shot 2018-06-18 at 11 40 22 am

File: src/Viz.js [hunk 5]
```diff
@@ -423,10 +424,24 @@ export default class Viz extends BaseClass {
...
+    // Updates the <title> tag if already exists else creates a new <title> tag on this.select.
+    const svgTitleText = this._title || "D3plus Visualization";
+    const svgTitle = this._select.selectAll("title").data([0]);
+    const svgTitleEnter = svgTitle.enter().append("title");
+    svgTitle.merge(svgTitleEnter).text(svgTitleText).attr("id", `${this._uuid}-title`);
+
+    // Updates the <desc> tag if already exists else creates a new <desc> tag on this.select.
+    const svgDesc = this._select.selectAll("desc").data([0]);
+    const svgDescEnter = svgDesc.enter().append("desc");
+    svgDesc.merge(svgDescEnter).text(this._svgDesc).attr("id", `${this._uuid}-desc`);
```
Comment (davelandry) [inline src/Viz.js]: Great job with these enter/update functions. One small tweak that would make this more precise: setting the `"id"` attribute only needs to occur on enter, as the uuid does not change on subsequent renders. You can move the `.attr("id", `${this._uuid}-desc`)` code for both the title and desc up 1 line to where you append the item.

File: src/Viz.js [hunk 6]
```diff
@@ -337,11 +342,9 @@ export default class Viz extends BaseClass {
     if (this.legendPosition() === "top" || this.legendPosition() === "bottom") drawLegend.bind(this)(this._filteredData);
     if (this.colorScalePosition() === "top" || this.colorScalePosition() === "bottom") drawColorScale.bind(this)(this._filteredData);
 
-    this._shapes = [];
```
Comment (davelandry) [inline src/Viz.js]: sorry I missed this on the last review: we can't delete this line, as the `this._shapes` variable needs to be reset to empty on each redraw. 

additionally, please see my previous comment about pulling from master to fix the merge conflict.

Comment (davelandry) [reply src/Viz.js]: @rbaheti I still need you to fix this issue.

Comment (rbaheti) [reply src/Viz.js]: Oh no! I thought I fixed it. :(

**Pull request merged**